base/src/types.rs

```diff
@@ -386,6 +289,45 @@ pub struct Field<Id, T = ArcType<Id>> {
     pub typ: T,
 }
 
+/// The representation of gluon's types.
+///
+/// For efficency this enum is not stored directly but instead a pointer wrapper which derefs to
+/// `Type` is used to enable types to be shared. It is recommended to use the static functions on
+/// `Type` such as `Type::app` and `Type::record` when constructing types as those will construct
+/// the pointer wrapper directly.
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub enum Type<Id, T = ArcType<Id>> {
+    /// An unbound type `_`, awaiting ascription.
+    Hole,
+    /// An application with multiple arguments.
+    /// `Map String Int` would be represented as `App(Map, [String, Int])`
+    App(T, Vec<T>),
+    /// A variant type `| A Int Float | B`.
+    /// The second element of the tuple is the function type which the constructor has which in the
+    /// above example means that A's type is `Int -> Float -> A` and B's is `B`
+    Variants(Vec<(Id, T)>),
+    /// Representation for type variables
+    Variable(TypeVariable),
+    /// Variant for "generic" variables. These occur in signatures as lowercase identifers `a`, `b`
+    /// etc and are what unbound type variables are eventually made into.
+    Generic(Generic<Id>),
+    /// A builtin type
+    Builtin(BuiltinType),
+    /// A record type
+    Record(T),
```

Comment:
Now you aren't even cleaning up legacy code, just my laziness/unawareness (after factoring out the types field) :stuck_out_tongue:.